File: web/packages/shared/components/AccessRequests/Shared/utils.test.ts
/**
 * Teleport
 * Copyright (C) 2026  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import {
  ResourceConstraints,
  ResourceIDString,
} from 'shared/services/accessRequests';

import {
  formatAWSRoleARNForDisplay,
  toggleAWSConsoleConstraint,
  toggleSSHConstraint,
} from './utils';

describe('formatAWSRoleARNForDisplay', () => {
  it('formats a valid ARN as "accountId: rolePathAndName"', () => {
    expect(
      formatAWSRoleARNForDisplay('arn:aws:iam::123456789012:role/Admin')
    ).toBe('123456789012: Admin');
  });

  it('handles path-based role names', () => {
    expect(
      formatAWSRoleARNForDisplay(
        'arn:aws:iam::123456789012:role/path/to/MyRole'
      )
    ).toBe('123456789012: path/to/MyRole');
  });

  it('returns the original string for non-matching ARNs', () => {
    expect(formatAWSRoleARNForDisplay('not-an-arn')).toBe('not-an-arn');
  });
});

describe('toggleAWSConsoleConstraint', () => {
  it('removes the specified ARN and keeps remaining', () => {
    const set = jest.fn();
    toggleAWSConsoleConstraint(
      {
        id: 'my-app',
        kind: 'app' as const,
        clusterName: 'cluster',
        constraints: {
          aws_console: {
            role_arns: ['arn:1', 'arn:2'],
          },
        },
      },
      'arn:1',
      set
    );
    expect(set).toHaveBeenCalledWith('cluster/app/my-app' as ResourceIDString, {
      aws_console: { role_arns: ['arn:2'] },
    });
  });

  it('clears constraints when last ARN is removed', () => {
    const set = jest.fn();
    toggleAWSConsoleConstraint(
      {
        id: 'my-app',
        kind: 'app' as const,
        clusterName: 'cluster',
        constraints: {
          aws_console: {
            role_arns: ['arn:1'],
          },
        },
      },
      'arn:1',
      set
    );
    expect(set).toHaveBeenCalledWith(
      'cluster/app/my-app' as ResourceIDString,
      undefined
    );
  });
});

describe('toggleSSHConstraint', () => {
  it('removes the specified login and keeps remaining', () => {
    const set = jest.fn<void, [ResourceIDString, ResourceConstraints?]>();
    toggleSSHConstraint(
      {
        id: 'my-node',
        kind: 'node' as const,
        clusterName: 'cluster',
        constraints: {
          ssh: { logins: ['root', 'ubuntu'] },
        },
      },
      'root',
      set
    );
    expect(set).toHaveBeenCalledWith(
      'cluster/node/my-node' as ResourceIDString,
      { ssh: { logins: ['ubuntu'] } }
    );
  });

  it('clears constraints when last login is removed', () => {
    const set = jest.fn<void, [ResourceIDString, ResourceConstraints?]>();
    toggleSSHConstraint(
      {
        id: 'my-node',
        kind: 'node' as const,
        clusterName: 'cluster',
        constraints: {
          ssh: { logins: ['root'] },
        },
      },
      'root',
      set
    );
    expect(set).toHaveBeenCalledWith(
      'cluster/node/my-node' as ResourceIDString,
      undefined
    );
  });
});

File: lib/scopes/cache/utils_test.go
/*
 * Teleport
 * Copyright (C) 2025  Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package cache

import (
	"testing"

	"github.com/stretchr/testify/require"
)

func TestStringCursorFormat(t *testing.T) {
	t.Parallel()

	tts := []struct {
		name   string
		cursor Cursor[string]
		enc    string
		ok     bool
	}{
		{
			name:   "empty",
			cursor: Cursor[string]{},
			enc:    "",
			ok:     true,
		},
		{
			name:   "root",
			cursor: Cursor[string]{Key: "foo", Scope: "/"},
			enc:    "v1:foo@/",
			ok:     true,
		},
		{
			name:   "non-root",
			cursor: Cursor[string]{Key: "foo", Scope: "/bar"},
			enc:    "v1:foo@/bar",
			ok:     true,
		},
		{
			name:   "empty scope",
			cursor: Cursor[string]{Key: "foo", Scope: ""},
			enc:    "v1:foo@",
			ok:     false,
		},
		{
			name:   "empty key",
			cursor: Cursor[string]{Key: "", Scope: "/bar"},
			enc:    "v1:@/bar",
			ok:     false,
		},
	}

	for _, tt := range tts {
		t.Run(tt.name, func(t *testing.T) {
			enc, err := EncodeStringCursor(tt.cursor)
			if tt.ok {
				require.NoError(t, err)
				require.Equal(t, tt.enc, enc)
			} else {
				require.Error(t, err)
				require.Empty(t, enc)
			}

			cursor, err := DecodeStringCursor(tt.enc)
			if tt.ok {
				require.NoError(t, err)
				require.Equal(t, tt.cursor, cursor)
			} else {
				require.Error(t, err)
				require.Equal(t, Cursor[string]{}, cursor)
			}
		})
	}
}
